### PR TITLE
strip, lstrip and rstrip operations implementation

### PR DESCRIPTION
**Pull request by shssf**
This PR is implements `strip`, `rstrip` and `lstrip` operations.
Tests are added into the group of similar operations: `strip`, `rstrip` and `lstrip`.
This PR is ready for review.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch. I've scheduled it for review.

File: numba/unicode.py [hunk 1]
```diff
@@ -696,6 +696,66 @@ def zfill_impl(string, width):
...
+        while right_it >= left_it and string[right_it] in chars:
+            right_it -= 1
+        right_it += 1
+
+    return string[left_it:right_it]
+
+
+@overload_method(types.UnicodeType, 'lstrip')
+def unicode_lstrip(string, chars=' '):
+    if not (chars == ' ' or isinstance(chars, (types.Omitted, types.UnicodeType, types.NoneType))):
```
Comment (esc) [inline numba/unicode.py]: Please place this type check in it's own function.

Comment (esc) [reply numba/unicode.py]: And then please reuse that function in the other two overloads.

File: numba/unicode.py [hunk 2]
```diff
@@ -696,6 +696,66 @@ def zfill_impl(string, width):
...
+        raise TypingError('The arg must be a UnicodeType or None')
+
+    def wrap(string, chars=' '):
+        return strip_impl(0, string, chars)
+    return wrap
+
+    return wrap
```
Comment (esc) [inline numba/unicode.py]: This `return` statement seems to be duplicated, also the case for the following two functions.

File: numba/unicode.py [hunk 3]
```diff
@@ -696,6 +696,66 @@ def zfill_impl(string, width):
...
+
+    if chars is None:
+        chars = ' '
+
+    str_len = len(string)
+
+    left_it = 0
+    right_it = str_len
+
+    if func_id != 1:
```
Comment (esc) [inline numba/unicode.py]: Here, I would advocate for some global, const style values, and checking for a positive match, something like:

```
if func_id in [LSTRIP, STRIP]:
    ...
```

**Comment by esc** [comment]
@shssf thanks for the patch, I've completed an initial review and made a few suggestions.

File: numba/unicode.py (hunk 3 above)
Comment (shssf) [reply numba/unicode.py]: @esc I tried to address this. If you meant something else please let me know. Perhaps I need to implement this http://code.activestate.com/recipes/65207-constants-in-python/?in=user-97991
If you require me to implement `class _const` as described in the link - I would propose to make it in some common place (please recommend a file).

File: numba/unicode.py [hunk 4]
```diff
@@ -696,6 +696,68 @@ def zfill_impl(string, width):
...
+def unicode_lstrip(string, chars=' '):
+    unicode_lstrip_types_check(chars)
+
+    def wrap(string, chars=' '):
+        return strip_impl(UNICODE_OP_LSTRIP, string, chars)
+    return wrap
+
+
+@overload_method(types.UnicodeType, 'rstrip')
+def unicode_rtrip(string, chars=' '):
```
Comment (esc) [inline numba/unicode.py]: There is a typo here, should be `rstrip`.

File: numba/unicode.py (hunk 3 above)
Comment (esc) [reply numba/unicode.py]: No, this is what I had in mind.

File: numba/unicode.py [hunk 5]
```diff
@@ -696,6 +696,68 @@ def zfill_impl(string, width):
...
 
 
+# const style variable to enumerate variants of "strip" operations family
+UNICODE_OP_LSTRIP = 0
+UNICODE_OP_RSTRIP = 1
+UNICODE_OP_STRIP = 2
+
+
+@register_jitable
+def strip_impl(func_id, string, chars=' '):
```
Comment (esc) [inline numba/unicode.py]: Default value for `chars` should be `None`.

Comment (shssf) [reply numba/unicode.py]: @esc for this line only or for all lines which contains `chars=' '` in this PR?

**Comment by esc** [comment]
@shssf I thought about this some more yesterday, and I think this will need some more work. I would propose to replace the runtime switches (the constants) with a set of composable functions. Let me sketch out what I had in mind, pseudocode follows:

```
def check_type(chars):
    # check the type of chars and raise if mismatch

def left_bound(string, chars):
    for i in range(len(string)):
        if string[i] not in chars:
            return i
    else:
        return len(string)

def right_bound(string, chars):
    # as above but for right

@overload_method
def unicode_lstrip(string, chars):
    check_type(chars)
    def lstrip_impl(string, chars):
        return string[left_bound(string, chars):]
```

Also, you could consider turning `chars` into a typed dictionary for speed and so you don't incur the `O(n)` penalty every time you check if a `char` is in `chars`. It might only make sense for larger `chars` but only profiling/timing will tell.

BTW: I am also available on the Numba channel on gitter, in case you want to discuss in real-time.

File: numba/unicode.py (hunk 5 above)
Comment (esc) [reply numba/unicode.py]: Probably for all lines, for consistency.

**Comment by shssf** [comment]
@esc I tried to address your comments except converting `chars` into dictionary.
Also, you might want me to move
```
    chars = ' ' if chars is None else chars
    str_len = len(string)
```
into `unicode_strip_types_check()`

**Comment by esc** [comment]
@shssf no, it turns out that moving that won't work because it needs to happen in the implementation rather than in the typing phase.

**Comment by esc** [comment]
@shssf it looks pretty good sofar! Let's wait for CI to complete.

**Comment by esc** [comment]
@shssf I just notices, you may need to use OptionalTypes here:

http://numba.pydata.org/numba-doc/latest/reference/types.html?highlight=optional#optional-types

**Comment by shssf** [comment]
@esc Since all tests are passed, could you please change label on this PR?

**Comment by esc** [comment]
@shssf apologies if I wasn't clear enough about this. From what I can tell, this function will allow None to be passed in. From what I understand, to do this correctly, you will need to use optional types:

http://numba.pydata.org/numba-doc/latest/reference/types.html?highlight=optional#optional-types

I'll see if I can dig out an example for you, but feel free to grep the code-base in the meantime.

**Comment by stuartarchibald** [comment]
@shssf I think the use case that @esc is alluding to is like:
```python
from numba import njit

@njit
def foo(x):
    opt = 'z'
    if len(x) > 0:
        opt = None
    return x.strip(opt)

string = 'a'
assert foo(string) == foo.py_func(string)
```
which is not handled by the current implementation. In the code above `opt` is typed as an `OptionalType` (essentially a type union of `"None or <some other type>"`). As the `.*strip()` methods all have a default value of `chars=None` the code above is a perfectly acceptable use but is not handled. I think checking if the type of `chars` is an instance of `types.Optional` and then ensuring that if it is the `type` member is an instance of `types.Unicode` would work.

**Comment by shssf** [comment]
@esc Is this change expected?
```
def unicode_strip_types_check(chars):

    chars = chars.type if isinstance(chars, types.Optional) else chars

    if not (chars is None or isinstance(chars, (types.Omitted, types.UnicodeType, types.NoneType))):
        raise TypingError('The arg must be a UnicodeType or None')
```

**Comment by stuartarchibald** [comment]
@shssf that looks like the change I'd expect.

**Comment by esc** [comment]
This looks good, @stuartarchibald shall we mark it for merge?

**Comment by stuartarchibald** [comment]
@esc Perhaps it might be a good idea to test optional types do indeed work?

**Comment by esc** [comment]
@stuartarchibald yes, that makes sense, @shssf can you take care of it?

**Comment by stuartarchibald** [comment]
@shssf thanks for all your work on this. With view of getting it merged in time for the 0.44 release candidate I've pushed the missing tests for optional types onto your branch. Hope this is ok?

@esc please could you check the diff from 6cd0866 to `HEAD` and make sure you are happy with the changes. Thanks.

**Comment by shssf** [comment]
@stuartarchibald Thank you

**Comment by stuartarchibald** [comment]
@shssf no problem :)

**Comment by seibert** [comment]
This looks ready to merge.

**Pull request merged**